### ISSUE TITLE
Calls to getDiagnostics in SDK 1.22 fail (due to new driver being enabled by default)

### ISSUE THREAD

**Issue by sethladd** [Dart-Code/Dart-Code#244]
Please review the below report for any information you do not wish to share and report to
  https://github.com/dart-lang/sdk/issues/new

Exception from analysis server (running from VSCode / Dart Code)

### What I was doing

(please describe what you were doing when this exception occurred)

### Request

While responding to request: `diagnostic.getDiagnostics`

### Versions

- Dart SDK 1.22.0-dev.5.0
- Visual Studio Code 1.8.1
- Dart Code 0.14.7

### Analyzer Info

The analyzer was launched with the arguments:

```text
--client-id=DanTup.dart-code
--client-version=0.14.7
```

### Exception

Bad state: Should not be used with the new analysis driver

```text
#0      ContextManagerImpl.folderMap (package:analysis_server/src/context_manager.dart:561)
#1      ContextManagerImpl.analysisContexts (package:analysis_server/src/context_manager.dart:557)
#2      AnalysisServer.analysisContexts (package:analysis_server/src/analysis_server.dart:447)
#3      DiagnosticDomainHandler.computeDiagnostics (package:analysis_server/src/domain_diagnostic.dart:42)
#4      DiagnosticDomainHandler.handleRequest (package:analysis_server/src/domain_diagnostic.dart:92)
#5      AnalysisServer.handleRequest.<anonymous closure>.<anonymous closure> (package:analysis_server/src/analysis_server.dart:876)
#6      _PerformanceTagImpl.makeCurrentWhile (package:analyzer/src/generated/utilities_general.dart:189)
#7      AnalysisServer.handleRequest.<anonymous closure> (package:analysis_server/src/analysis_server.dart:872)
#8      _rootRun (dart:async/zone.dart:1150)
#9      _CustomZone.run (dart:async/zone.dart:1026)
#10     _CustomZone.runGuarded (dart:async/zone.dart:924)
#11     runZoned (dart:async/zone.dart:1501)
#12     AnalysisServer.handleRequest (package:analysis_server/src/analysis_server.dart:871)
#13     ByteStreamServerChannel._readRequest.<anonymous closure> (package:analysis_server/src/channel/byte_stream_channel.dart:170)
#14     _PerformanceTagImpl.makeCurrentWhile (package:analyzer/src/generated/utilities_general.dart:189)
#15     ByteStreamServerChannel._readRequest (package:analysis_server/src/channel/byte_stream_channel.dart:161)
#16     ByteStreamServerChannel.listen.<anonymous closure> (package:analysis_server/src/channel/byte_stream_channel.dart:110)
#17     _rootRunUnary (dart:async/zone.dart:1158)
#18     _CustomZone.runUnary (dart:async/zone.dart:1037)
#19     _CustomZone.runUnaryGuarded (dart:async/zone.dart:932)
#20     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:331)
#21     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:258)
#22     _SinkTransformerStreamSubscription._add (dart:async/stream_transformers.dart:67)
#23     _EventSinkWrapper.add (dart:async/stream_transformers.dart:14)
#24     _StringAdapterSink.add (dart:convert/string_conversion.dart:256)
#25     _LineSplitterSink._addLines (dart:convert/line_splitter.dart:159)
#26     _LineSplitterSink.addSlice (dart:convert/line_splitter.dart:134)
#27     StringConversionSinkMixin.add (dart:convert/string_conversion.dart:180)
#28     _SinkTransformerStreamSubscription._handleData (dart:async/stream_transformers.dart:119)
#29     _rootRunUnary (dart:async/zone.dart:1158)
#30     _CustomZone.runUnary (dart:async/zone.dart:1037)
#31     _CustomZone.runUnaryGuarded (dart:async/zone.dart:932)
#32     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:331)
#33     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:258)
#34     _SinkTransformerStreamSubscription._add (dart:async/stream_transformers.dart:67)
#35     _EventSinkWrapper.add (dart:async/stream_transformers.dart:14)
#36     _StringAdapterSink.add (dart:convert/string_conversion.dart:256)
#37     _StringAdapterSink.addSlice (dart:convert/string_conversion.dart:260)
#38     _Utf8ConversionSink.addSlice (dart:convert/string_conversion.dart:336)
#39     _Utf8ConversionSink.add (dart:convert/string_conversion.dart:329)
#40     _ConverterStreamEventSink.add (dart:convert/chunked_conversion.dart:92)
#41     _SinkTransformerStreamSubscription._handleData (dart:async/stream_transformers.dart:119)
#42     _rootRunUnary (dart:async/zone.dart:1158)
#43     _CustomZone.runUnary (dart:async/zone.dart:1037)
#44     _CustomZone.runUnaryGuarded (dart:async/zone.dart:932)
#45     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:331)
#46     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:258)
#47     _StreamController&&_SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:768)
#48     _StreamController._add (dart:async/stream_controller.dart:635)
#49     _StreamController.add (dart:async/stream_controller.dart:581)
#50     _Socket._onData (dart:io-patch/socket_patch.dart:1680)
#51     _rootRunUnary (dart:async/zone.dart:1162)
#52     _CustomZone.runUnary (dart:async/zone.dart:1037)
#53     _CustomZone.runUnaryGuarded (dart:async/zone.dart:932)
#54     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:331)
#55     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:258)
#56     _StreamController&&_SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:768)
#57     _StreamController._add (dart:async/stream_controller.dart:635)
#58     _StreamController.add (dart:async/stream_controller.dart:581)
#59     _RawSocket._RawSocket.<anonymous closure> (dart:io-patch/socket_patch.dart:1247)
#60     _NativeSocket.issueReadEvent.issue (dart:io-patch/socket_patch.dart:781)
#61     _microtaskLoop (dart:async/schedule_microtask.dart:41)
#62     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50)
#63     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:96)
#64     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:149)
```

**Comment by DanTup** [comment]
Hmmm, this is confusing. The `getDiagnostics` call does not work with the new driver (this is something in-development in the analysis server that must be turned on with a flag). I added a change in commit 589b8939a8fd5508d1b518afde78179846450e35 so that we wouldn't try to call it when using the new driver.

However, in your log it does not show that you're passing the flag to enable the new driver!

@devoncarew @bwilkerson Is the new driver enabled by default in `SDK 1.22.0-dev.5.0`; should I take the `getDiagnostics` call out for now (or at least, detect >= `1.22` instead of just the `--enable-new-analysis-driver` switch being passed)?

**Comment by DanTup** [comment]
Found https://github.com/dart-lang/sdk/issues/28244. Looks like new driver is enabled by default now; my method of detecting it is therefore flawed!

@sethladd I'll push an update as soon as I can (probably this weekend) to disable this (at least until the new driver has some equiv function).

For now, you might be able to fix this by adding `--enable-new-analysis-driver` to the `dart.analyzerAdditionalArgs` setting which will inform Dart Code you're using the new driver and suppress this call (as long as the analyzer doesn't barf on this switch because it's now redundant).

Failing that, you might have to switch back to an SDK version prior to this change until I've pushed a fix. Sorry for the inconvenience; it didn't (but should've) occur to me that the new driver might become default without the getDiagnostics call implemented.

**Comment by bwilkerson** [comment]
Yes, the new driver is enabled by default now. I believe (though I haven't confirmed) that server should ignore any flags it doesn't recognize, so it should be safe to add the flag.

Here's the issue tracking the problem with the server API: https://github.com/dart-lang/sdk/issues/28068.

**Comment by DanTup** [comment]
@bwilkerson Great, thanks for confirming! @sethladd in that case, adding `--enable-new-analysis-driver` as mentioned above should solve the issue until I can push a fix out.

**Comment by sethladd** [comment]
Screenshot, just for the record:

![screen shot 2017-01-06 at 2 05 30 pm](https://cloud.githubusercontent.com/assets/5479/21734645/47af083a-d419-11e6-9d5e-62fb95151a8c.png)

**Comment by DanTup** [comment]
@sethladd I've pushed out a fix (commit ccc1ce55223abb8fe27b80b7f4082e82c4a45f22) in v0.14.8 that stops this (there are no other changes).

Note: The other error in your screenshot (second one) appears to be from the analysis server and should be raised in https://github.com/dart-lang/sdk. Once you've updated to v0.14.8 you should be able to generate an error report for it without getting the other error.

**Comment by sethladd** [comment]
Thanks! I'll give it a shot.
